From 1d259bf1d4f627609ac0e959db150d0886a0a596 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 26 May 2022 07:06:48 +0000
Subject: [PATCH 2659/4585] [Huawei] mm/sharepool: Add a task_struct parameter
 for sp_get_local_group()

Offering: HULK
ascend inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I5DS9S

-------------------------------------------------

sp_get_local_group() could be invoked in kthread, where the current
process isn't the process we want. Add a parameter and let the caller to
avoid this problem.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 59db0e27880b..0f1597e37987 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -367,7 +367,7 @@ static inline bool is_local_group(int spg_id)
 	return spg_id >= SPG_ID_LOCAL_MIN && spg_id <= SPG_ID_LOCAL_MAX;
 }
 
-static struct sp_group *sp_get_local_group(struct mm_struct *mm)
+static struct sp_group *sp_get_local_group(struct task_struct *tsk, struct mm_struct *mm)
 {
 	int ret;
 	struct sp_group_master *master;
@@ -382,7 +382,7 @@ static struct sp_group *sp_get_local_group(struct mm_struct *mm)
 	up_read(&sp_group_sem);
 
 	down_write(&sp_group_sem);
-	ret = sp_init_group_master_locked(current, mm);
+	ret = sp_init_group_master_locked(tsk, mm);
 	if (ret) {
 		up_write(&sp_group_sem);
 		return ERR_PTR(ret);
@@ -2513,7 +2513,7 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 		}
 		ac->type = SPA_TYPE_ALLOC;
 	} else {  /* allocation pass through scene */
-		spg = sp_get_local_group(current->mm);
+		spg = sp_get_local_group(current, current->mm);
 		if (IS_ERR(spg))
 			return PTR_ERR(spg);
 		down_read(&spg->rw_lock);
@@ -3882,7 +3882,7 @@ bool sp_config_dvpp_range(size_t start, size_t size, int device_id, int pid)
 	if (!mm)
 		goto put_task;
 
-	spg = sp_get_local_group(mm);
+	spg = sp_get_local_group(tsk, mm);
 	if (IS_ERR(spg))
 		goto put_mm;
 
-- 
2.17.1

